README.md
# factorio-factory-levels
A mod for factorio [Link to mod portal](https://mods.factorio.com/mod/factory-levels-fork)

You want Assemblers and Furnaces that upgrade with more Items produced and smelted?
This is the mod for you.


factory-levels/changelog.txt
THIS IS A FORK OF VERSION 0.6.8
Version: 1.0.0
Date: 2023-01-27
  Changes:
   - Added "Electric Furnaces" support.
   Bugfix:
   - Fixed bug when liquifier from Angels Refining upgraded to ore crusher



---------------------------------------------------------------------------------------------------
Version: 0.6.8
Date: 2022-09-24
  Changes:
    - Thanks to CaitSith2: mods via remote interface can now declare they wish to be disabled.
    - Thanks to CaitSith2: machines no longer instantly level up to the next tier the moment they hit max level of current tier.
    - Added support for Algae Farm from Angels Bioprocessing
  Bugfix:
    - Thanks to CaitSith2: fixed a bug with on_configuration_changed. (this event fires on updating any mods, or installing/removing any mods.)
---------------------------------------------------------------------------------------------------
Version: 0.6.7
Date: 2022-09-07
  Bugfix:
    - Thanks to CaitSith2: Fixed a crash when copy/pasting a leveled machine with modules in editor mode with instant blueprinting turned on.
---------------------------------------------------------------------------------------------------
Version: 0.6.6
Date: 2022-09-06
  Features:
    - Added support for Angel's Refining and Angel's Petrochemical Processing
---------------------------------------------------------------------------------------------------
Version: 0.6.5
Date: 2022-09-05
  Changes:
    - Thanks to CaitSith2: Changelog now shows up properly in factorio.
    - Thanks to CaitSith2: Updated changelog for changes made in version 0.6.1 to 0.6.4
---------------------------------------------------------------------------------------------------
Version: 0.6.4
Date: 2022-09-05
  Bugfix:
    - Thanks to CaitSith2: Fix possibility of referring to a nil value.
---------------------------------------------------------------------------------------------------
Version: 0.6.3
Date: 2022-09-03
  Changes:
    - Thanks to CaitSith2: Localised name now automatically derived from formatted string of  ('Machine_base_name' Level 'Level')
  Bugfix:
    - Thanks to CaitSith2: Localised description now references the base machine.
    - Thanks to CaitSith2: Fix Invalid key for 'next' error.
---------------------------------------------------------------------------------------------------
Version: 0.6.2
Date: 2022-09-03
  Features:
    - Thanks to CaitSith2: Add setting that allows for safe removal of mod.
  Changes:
    - Thanks to CaitSith2: Iterate over known machines instead of chunks.
  Bugfix:
    - Thanks to CaitSith2: Machines on surfaces created since installation of mod now correctly level up.
---------------------------------------------------------------------------------------------------
Version: 0.6.1
Date: 2022-08-23
  Bugfix:
    - Thanks to CaitSith2: Inputs and outputs no longer voided on level up.
    - Thanks to CaitSith2: Fixed regression of SE machine level productivity.
---------------------------------------------------------------------------------------------------
Version: 0.6.0
Date: 2022-08-15
  Bugfix:
    - Thanks to CaitSith2: Chemical plant, refinery, electric smelter and centrifuge
    - Thanks to CaitSith2: refactoring of code
---------------------------------------------------------------------------------------------------
Version: 0.5.2
Date: 2022-08-14
  Changes:
    - Disable refinery and chemical plant for now (modules are removed)
  Bugfix:
    - fix bug with ghost buildings?
---------------------------------------------------------------------------------------------------
Version: 0.5.1
Date: 2022-08-12
  Features:
    - Added levels for oil refinery and chemical plant
---------------------------------------------------------------------------------------------------
Version: 0.5.0
Date: 2022-08-12
  Features:
    - Thanks to CaitSith2: Added remote interface command update_machine
    - Thanks to CaitSith2: Added remote interface command remove_machine
    - Thanks to CaitSith2: Added remote interface command get_machine
    - Thanks to CaitSith2: Machine level exponent is now a map setting.
---------------------------------------------------------------------------------------------------
Version: 0.4.2
Date: 2022-08-02
  Bugfix:
    - Space Exploration did not work with assembler productivity, now it does
---------------------------------------------------------------------------------------------------
Version: 0.4.1
Date: 2022-08-01
  Bugfix:
    - Fix initial crafting counts from testing
---------------------------------------------------------------------------------------------------
Version: 0.4.0
Date: 2022-08-01
  Changes:
    - Thanks to idea from unhott in alien-module: increased Performance of mod by iterating over chunks instead of updating everything at once
---------------------------------------------------------------------------------------------------
Version: 0.3.2
Date: 2022-07-28
  Changes:
    - Thanks to CaitSith2: Leveled Entities are now hidden in "made in" Tooltip
---------------------------------------------------------------------------------------------------
Version: 0.3.1
Date: 2022-07-26
  Changes:
    - Thanks to CaitSith2: Refactoring of Code
  Bugfix:
    - Thanks to CaitSith2: Construction via bots also works now
---------------------------------------------------------------------------------------------------
Version: 0.3.0
Date: 2022-07-25
  Features:
    - Thanks to CaitSith2: External API for adding entities to leveling
---------------------------------------------------------------------------------------------------
Version: 0.2.5
Date: 2022-07-24
  Changes:
    - Thanks to CaitSith2: Refactor machine level code.
  Bugfix:
    - Spilling of input is fixed now
---------------------------------------------------------------------------------------------------

factory-levels-fork_1.0.0/settings.lua
data:extend({
    {
        type = "bool-setting",
        name = "factory-levels-enable-productivity-bonus",
        order = "a",
        setting_type = "startup",
        default_value = true
    },
    {
        type = "bool-setting",
        name = "factory-levels-enable-module-bonus",
        order = "aa",
        setting_type = "startup",
        default_value = true
    },
    {
        type = "bool-setting",
        name = "factory-levels-enable-speed-bonus",
        order = "b",
        setting_type = "startup",
        default_value = true
    },
    {
        type = "bool-setting",
        name = "factory-levels-enable-energy-usage",
        order = "c",
        setting_type = "startup",
        default_value = true
    },
    {
        type = "bool-setting",
        name = "factory-levels-enable-emissions",
        order = "d",
        setting_type = "startup",
        default_value = true
    },
    {
        type = "double-setting",
        name = "factory-levels-exponent",
        order = "e",
        minimum_value = 1.5, maximum_value = 5,
        setting_type = "runtime-global",
        default_value = 3
    },
	{
		type = "bool-setting",
		name = "factory-levels-disable-mod",
		order = "f",
		setting_type = "runtime-global",
		default_value = false
	}
})

factory-levels-fork_1.0.0/prototypes/entity/factory_levels.lua
local factory_levels = {}

function factory_levels.update_machine_speed(machine, level, base_speed, speed_multiplier)
	if (settings.startup["factory-levels-enable-speed-bonus"].value) then
		machine.crafting_speed = base_speed + level * speed_multiplier
	end
end

function factory_levels.update_machine_energy_usage(machine, level, base_usage, usage_multiplier, energy_unit)
	if (settings.startup["factory-levels-enable-energy-usage"].value) then
		machine.energy_usage = (base_usage + usage_multiplier * level) .. energy_unit
	end
end

function factory_levels.update_machine_pollution(machine, level, base_emission, emission_multiplier)
	if (settings.startup["factory-levels-enable-emissions"].value) then
		machine.energy_source.emissions_per_minute = base_emission + emission_multiplier * level
	end
end

function factory_levels.update_machine_productivity(machine, level, base_productivity, productivity_multiplier)
	if (settings.startup["factory-levels-enable-productivity-bonus"].value) then
		machine.base_productivity = base_productivity + productivity_multiplier * level
	end
end

function factory_levels.update_machine_module_slots(machine, level, levels_per_module_slot, base_module_slots, module_slot_bonus)
	if (settings.startup["factory-levels-enable-module-bonus"].value) then
		machine.module_specification = { module_slots = base_module_slots + (math.floor(level / levels_per_module_slot) * module_slot_bonus) }
		if machine.module_specification.module_slots > 0 then
			machine.allowed_effects = { "consumption", "speed", "productivity", "pollution" }
		end
	end
end

function factory_levels.update_animation_tint(animation, tint)
	if animation == nil then
		return
	end

	if animation.north then
		factory_levels.update_animation_tint(animation.north, tint)
		factory_levels.update_animation_tint(animation.east, tint)
		factory_levels.update_animation_tint(animation.west, tint)
		factory_levels.update_animation_tint(animation.south, tint)
		return
	end

	if animation.layers then
		for _, layer in pairs(animation.layers) do
			factory_levels.update_animation_tint(layer, tint)
		end
	end
	if animation.hr_version then
		factory_levels.update_animation_tint(animation.hr_version, tint)
	end
	animation.tint = tint
end

function factory_levels.update_machine_tint(machine, level, base_tint, tint_multiplier)
	local tint = { r = base_tint.r + level * tint_multiplier.r,
				   g = base_tint.g + level * tint_multiplier.g,
				   b = base_tint.b + level * tint_multiplier.b,
				   a = 1 }

	factory_levels.update_animation_tint(machine.animation, tint)

	-- Some machines, such as centrigue put their main animation set into idle_animation.
	factory_levels.update_animation_tint(machine.idle_animation, tint)
end

function factory_levels.get_furnace(base_machine_name)
	-- Some mods might turn the stone-furnace, steel-furnace and electric-furnace into assembling machine prototype. Krastorio 2 being an example of one such mod.
	if data.raw["furnace"][base_machine_name] then
		return data.raw["furnace"][base_machine_name]
	end
	if data.raw["assembling-machine"][base_machine_name] then
		return data.raw["assembling-machine"][base_machine_name]
	end
	return nil
end

function factory_levels.get_or_create_machine(machine_type, base_machine_name, level)
	local base_machine = data.raw[machine_type][base_machine_name]

	if machine_type == "furnace" then
		base_machine = factory_levels.get_furnace(base_machine_name)
	end

	if base_machine == nil then
		return nil
	end

	if level == 0 then
		return base_machine
	end

	local new_machine_name = base_machine_name .. "-level-" .. level

	if data.raw[base_machine.type][new_machine_name] == nil then
		local base_machine = data.raw[machine_type][base_machine_name]
		if base_machine == nil then
			return nil
		end
		local machine = table.deepcopy(base_machine)
		machine.name = new_machine_name
		data:extend({ machine })
	end

	return data.raw[base_machine.type][new_machine_name]
end

function factory_levels.create_leveled_machines(machines)
	for tier = 1, machines.tiers, 1 do
		for level = 0, machines.levels[tier], 1 do
			local machine = factory_levels.get_or_create_machine(machines.type, machines.base_machine_names[tier], level)

			if level > 0 then
				machine.flags = machine.flags or { "placeable-neutral", "placeable-player", "player-creation" }
				local hidden = false

				for _, flag in pairs(machine.flags) do
					if flag == "hidden" then
						hidden = true
						break
					end
				end

				if not hidden then
					table.insert(machine.flags, "hidden")
				end

				machine.minable.result = machines.base_machine_names[tier]
				machine.placeable_by = { item = machines.base_machine_names[tier], count = 1 }
				machine.localised_name = { "entity-name.factory-levels", { "entity-name." .. machines.base_machine_names[tier] }, level }
				machine.localised_description = { "entity-description." .. machines.base_machine_names[tier] }
			end

			factory_levels.update_machine_speed(machine, level, machines.base_speeds[tier], machines.speed_multipliers[tier])
			factory_levels.update_machine_energy_usage(machine, level, machines.base_consumption[tier], machines.consumption_multipliers[tier], machines.consumption_unit[tier])
			factory_levels.update_machine_pollution(machine, level, machines.base_pollution[tier], machines.pollution_multipliers[tier])
			factory_levels.update_machine_productivity(machine, level, machines.base_productivity[tier], machines.productivity_multipliers[tier])
			factory_levels.update_machine_module_slots(machine, level, machines.levels_per_module_slots[tier], machines.base_module_slots[tier], machines.bonus_module_slots[tier])
			factory_levels.update_machine_tint(machine, level, machines.base_level_tints[tier], machines.level_tint_multipliers[tier])
		end
	end
end

-- Data final fixes
function factory_levels.convert_furnace_to_assembling_machines(machines)
	-- Just in case the base machine was initially "furnace" when the leveled machines were created, but another mod later turned it into "assembling-machine" without
	-- also converting the leveled machines into "assembling-machines".
	if machines.type ~= "furnace" then
		return
	end
	for tier = 1, machines.tiers, 1 do
		base_machine = factory_levels.get_or_create_machine(machines.type, machines.base_machine_names[tier], 0)
		if base_machine.type == "assembling-machine" then
			for level = 1, machines.levels[tier], 1 do
				leveled_machine = factory_levels.get_or_create_machine(machines.type, machines.base_machine_names[tier], level)
				if leveled_machine.type == "furnace" then
					data.raw["furnace"][leveled_machine.name] = nil
					leveled_machine.type = "assembling-machine"
					data:extend({ leveled_machine })
				end
			end
		end
	end
end

function factory_levels.fix_productivity(machines)
	if (settings.startup["factory-levels-enable-productivity-bonus"].value == false) then
		return
	end
	--Undo space-explorations hold against machine level productivity.
	for tier = 1, machines.tiers, 1 do
		for level = 0, machines.levels[tier], 1 do
			local machine = factory_levels.get_or_create_machine(machines.type, machines.base_machine_names[tier], level)
			if machine.base_productivity == 0 or machine.base_productivity == nil then
				factory_levels.update_machine_productivity(machine, level, machines.base_productivity[tier], machines.productivity_multipliers[tier])
			end
		end
	end
end

return factory_levels

factory-levels-fork_1.0.0/prototypes/entity/entity.lua
assembling_machine_levels = {
	type = "assembling-machine",
	tiers = 3,
	base_machine_names = { "assembling-machine-1", "assembling-machine-2", "assembling-machine-3" },
	base_level_tints = {
		{ r = 1, g = 1, b = 1 },
		{ r = 1, g = 1, b = 1 },
		{ r = 1, g = 1, b = 1 }
	},
	level_tint_multipliers = {
		{ r = -0.02, g = -0.02, b = 0 },
		{ r = 0, g = 0, b = -0.016 },
		{ r = 0, g = -0.008, b = -0.008 },
	},
	levels = { 25, 50, 100 },
	base_speeds = { 0.5, 0.75, 1.25 },
	speed_multipliers = { 0.01, 0.01, 0.0175 },
	base_consumption = { 75, 150, 375 },
	consumption_multipliers = { 2.5, 4.5, 7.5 },
	consumption_unit = { "kW", "kW", "kW" },
	base_pollution = { 4, 3, 2, },
	pollution_multipliers = { 0.04, 0.04, 0.04 },
	base_productivity = { 0, 0, 0 },
	productivity_multipliers = { 0.0025, 0.0025, 0.0025 },
	levels_per_module_slots = { 20, 20, 25 },
	base_module_slots = { 0, 2, 4 },
	bonus_module_slots = { 1, 1, 1 }
}

factory_levels.create_leveled_machines(assembling_machine_levels)

oil_refinery_levels = {
	type = "assembling-machine",
	tiers = 1,
	base_machine_names = { "oil-refinery" },
	base_level_tints = { { r = 1, g = 1, b = 1 } },
	level_tint_multipliers = { { r = 0, g = -0.008, b = -0.008 } },
	levels = { 100 },
	base_speeds = { 1 },
	speed_multipliers = { 0.015 },
	base_consumption = { 210 },
	consumption_multipliers = { 5 },
	consumption_unit = { "kW" },
	base_pollution = { 6 },
	pollution_multipliers = { 0.14 },
	base_productivity = { 0 },
	productivity_multipliers = { 0.001 },
	levels_per_module_slots = { 25 },
	base_module_slots = { 3 },
	bonus_module_slots = { 1 }
}
factory_levels.create_leveled_machines(oil_refinery_levels)

chemical_plant_levels = {
	type = "assembling-machine",
	tiers = 1,
	base_machine_names = { "chemical-plant" },
	base_level_tints = { { r = 1, g = 1, b = 1 } },
	level_tint_multipliers = { { r = 0, g = -0.008, b = -0.008 } },
	levels = { 100 },
	base_speeds = { 1 },
	speed_multipliers = { 0.015 },
	base_consumption = { 210 },
	consumption_multipliers = { 5 },
	consumption_unit = { "kW" },
	base_pollution = { 4 },
	pollution_multipliers = { 0.06 },
	base_productivity = { 0 },
	productivity_multipliers = { 0.001 },
	levels_per_module_slots = { 25 },
	base_module_slots = { 3 },
	bonus_module_slots = { 1 }
}
if mods["space-exploration"] then
	table.insert(data.raw["assembling-machine"]["chemical-plant"].crafting_categories, "melting")
end
factory_levels.create_leveled_machines(chemical_plant_levels)

centrifuge_levels = {
	type = "assembling-machine",
	tiers = 1,
	base_machine_names = { "centrifuge" },
	base_level_tints = { { r = 1, g = 1, b = 1 } },
	level_tint_multipliers = { { r = 0, g = -0.008, b = -0.008 } },
	levels = { 100 },
	base_speeds = { 1 },
	speed_multipliers = { 0.015 },
	base_consumption = { 350 },
	consumption_multipliers = { 5 },
	consumption_unit = { "kW" },
	base_pollution = { 4 },
	pollution_multipliers = { 0.06 },
	base_productivity = { 0 },
	productivity_multipliers = { 0.001 },
	levels_per_module_slots = { 25 },
	base_module_slots = { 2 },
	bonus_module_slots = { 1 }
}
factory_levels.create_leveled_machines(centrifuge_levels)

-- Furnaces
burner_furnace_levels = {
	type = "furnace",
	tiers = 2,
	base_machine_names = { "stone-furnace", "steel-furnace" },
	base_level_tints = {
		{ r = 1, g = 1, b = 1 },
		{ r = 1, g = 1, b = 1 }
	},
	level_tint_multipliers = {
		{ r = 0, g = -0.032, b = -0.032 },
		{ r = 0, g = -0.008, b = -0.008 }
	},
	levels = { 25, 100 },
	base_speeds = { 1, 2 },
	speed_multipliers = { 0.04, 0.04 },
	base_consumption = { 90, 90 },
	consumption_multipliers = { 1, 2 },
	consumption_unit = { "kW", "kW" },
	base_pollution = { 2, 4 },
	pollution_multipliers = { 0.04, 0.12 },
	base_productivity = { 0, 0 },
	productivity_multipliers = { 0.002, 0.002 },
	levels_per_module_slots = { 25, 25 },
	base_module_slots = { 0, 0 },
	bonus_module_slots = { 0, 1 },
}

electric_furnace_levels = {
	type = "furnace",
	tiers = 1,
	base_machine_names = { "electric-furnace" },
	base_level_tints = { { r = 1, g = 1, b = 1 } },
	level_tint_multipliers = { { r = 0, g = -0.008, b = -0.008 } },
	levels = { 100 },
	base_speeds = { 2 },
	speed_multipliers = { 0.04 },
	base_consumption = { 180 },
	consumption_multipliers = { 2.8 },
	consumption_unit = { "kW" },
	base_pollution = { 1 },
	pollution_multipliers = { 0.12 },
	base_productivity = { 0 },
	productivity_multipliers = { 0.002 },
	levels_per_module_slots = { 25 },
	base_module_slots = { 2 },
	bonus_module_slots = { 1 }
}

if mods["space-exploration"] then
	table.insert(data.raw["furnace"]["stone-furnace"].crafting_categories, "kiln")
	table.insert(data.raw["furnace"]["steel-furnace"].crafting_categories, "kiln")
	table.insert(data.raw["furnace"]["electric-furnace"].crafting_categories, "kiln")
end

factory_levels.create_leveled_machines(burner_furnace_levels)
factory_levels.create_leveled_machines(electric_furnace_levels)

if mods ["Electric Furnaces"] then

electric_burner_furnace_levels = {
	type = "furnace",
	tiers = 2,
	base_machine_names = { "electric-stone-furnace", "electric-steel-furnace" },
	base_level_tints = {
		{ r = 1, g = 1, b = 1 },
		{ r = 1, g = 1, b = 1 }
	},
	level_tint_multipliers = {
		{ r = 0, g = -0.032, b = -0.032 },
		{ r = 0, g = -0.008, b = -0.008 }
	},
	levels = { 25, 100 },
	base_speeds = { 1, 2 },
	speed_multipliers = { 0.04, 0.04 },
	base_consumption = { 90, 90 },
	consumption_multipliers = { 1, 2 },
	consumption_unit = { "kW", "kW" },
	base_pollution = { 2, 4 },
	pollution_multipliers = { 0.04, 0.12 },
	base_productivity = { 0, 0 },
	productivity_multipliers = { 0.002, 0.002 },
	levels_per_module_slots = { 25, 25 },
	base_module_slots = { 0, 0 },
	bonus_module_slots = { 0, 1 },
}

electric_leveled_furnace_levels = {
	type = "furnace",
	tiers = 3,
	base_machine_names = { "electric-furnace", "electric-furnace-2", "electric-furnace-3" },
	base_level_tints = { { r = 1, g = 1, b = 1 },  { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
	level_tint_multipliers = { { r = 0, g = -0.008, b = -0.008 }, { r = 0, g = -0.008, b = -0.008 }, { r = 0, g = -0.008, b = -0.008 } },
	levels = { 100, 100, 100 },
	base_speeds = { 2, 2.5, 3 },
	speed_multipliers = { 0.04, 0.06, 0.08 },
	base_consumption = { 180, 240, 300 },
	consumption_multipliers = { 2.8, 2.8, 2.8 },
	consumption_unit = { "kW", "kW", "kW" },
	base_pollution = { 1, 1.1, 1.2 },
	pollution_multipliers = { 0.12, 0.12, 0.12 },
	base_productivity = { 0, 0, 0 },
	productivity_multipliers = { 0.002, 0.002, 0.002 },
	levels_per_module_slots = { 25, 25, 25 },
	base_module_slots = { 2, 2, 2 },
	bonus_module_slots = { 1, 1, 1 }
}
factory_levels.create_leveled_machines(electric_burner_furnace_levels)
factory_levels.create_leveled_machines(electric_leveled_furnace_levels)


end


if mods["angelspetrochem"] then
	electrolyzer_levels = {
		type = "assembling-machine",
		tiers = 4,
		base_machine_names = { "angels-electrolyser", "angels-electrolyser-2", "angels-electrolyser-3", "angels-electrolyser-4" },
		base_level_tints = { { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
		level_tint_multipliers = { { r = 0, g = 0, b = -0.02 }, { r = 0, g = 0, b = -0.02 }, { r = 0, g = -0.015, b = -0.015 }, { r = -0.009, g = 0, b = 0 } },
		levels = { 25, 25, 25, 100 },
		base_speeds = { 1, 1.5, 2, 2.5 },
		speed_multipliers = { 0.02, 0.02, 0.02, 0.025 },
		base_consumption = { 300, 350, 400, 450 },
		consumption_multipliers = { 2, 2, 2, 2 },
		consumption_unit = { "kW", "kW", "kW", "kW" },
		base_pollution = { 1.2, 1.8, 2.4, 3 },
		pollution_multipliers = { 0.1, 0.1, 0.1, 0.2 },
		base_productivity = { 0, 0, 0, 0 },
		productivity_multipliers = { 0.002, 0.002, 0.002, 0.002 },
		levels_per_module_slots = { 25, 25, 25, 25 },
		base_module_slots = { 2, 2, 2, 2 },
		bonus_module_slots = { 0, 0, 0, 1 }
	}

	factory_levels.create_leveled_machines(electrolyzer_levels)
end

if mods["angelsbioprocessing"] then
	algaefarm_levels = {
		type = "assembling-machine",
		tiers = 4,
		base_machine_names = { "algae-farm", "algae-farm-2", "algae-farm-3", "algae-farm-4" },
		base_level_tints = { { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
		level_tint_multipliers = { { r = 0, g = 0, b = -0.02 }, { r = 0, g = 0, b = -0.02 }, { r = 0, g = -0.015, b = -0.015 }, { r = -0.009, g = 0, b = 0 } },
		levels = { 25, 25, 25, 100 },
		base_speeds = { 0.5, 1, 1.5, 2 },
		speed_multipliers = { 0.02, 0.02, 0.02, 0.03 },
		base_consumption = { 100, 125, 150, 175 },
		consumption_multipliers = { 1, 1, 1, 1 },
		consumption_unit = { "kW", "kW", "kW", "kW" },
		base_pollution = { -10, -20, -30, -40 },
		pollution_multipliers = { -0.1, -0.1, -0.1, -0.2 },
		base_productivity = { 0, 0, 0, 0 },
		productivity_multipliers = { 0.002, 0.002, 0.002, 0.002 },
		levels_per_module_slots = { 25, 25, 25, 25 },
		base_module_slots = { 2, 2, 2, 2 },
		bonus_module_slots = { 0, 0, 0, 1 }
	}

	factory_levels.create_leveled_machines(algaefarm_levels)
end

if mods["angelsrefining"] then
	ore_crusher_levels = {
		type = "assembling-machine",
		tiers = 4,
		base_machine_names = { "burner-ore-crusher", "ore-crusher", "ore-crusher-2", "ore-crusher-3" },
		base_level_tints = { { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
		level_tint_multipliers = { { r = 0, g = 0, b = -0.02 }, { r = 0, g = 0, b = -0.02 }, { r = 0, g = -0.015, b = -0.015 }, { r = -0.009, g = 0, b = 0 } },
		levels = { 25, 25, 50, 100 },
		base_speeds = { 1, 1.5, 2, 3 },
		speed_multipliers = { 0.02, 0.02, 0.02, 0.04 },
		base_consumption = { 100, 100, 125, 150 },
		consumption_multipliers = { 2, 2, 2, 2 },
		consumption_unit = { "kW", "kW", "kW", "kW" },
		base_pollution = { 7, 2, 3, 4 },
		pollution_multipliers = { 0.1, 0.1, 0.1, 0.2 },
		base_productivity = { 0, 0, 0, 0 },
		productivity_multipliers = { 0.002, 0.002, 0.002, 0.002 },
		levels_per_module_slots = { 25, 25, 25, 25 },
		base_module_slots = { 0, 1, 2, 3 },
		bonus_module_slots = { 0, 0, 1, 1 }
	}

	liquifier_levels = {
		type = "assembling-machine",
		tiers = 4,
		base_machine_names = { "liquifier", "liquifier-2", "liquifier-3", "liquifier-4" },
		base_level_tints = { { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
		level_tint_multipliers = { { r = 0, g = 0, b = -0.02 }, { r = 0, g = 0, b = -0.02 }, { r = 0, g = -0.015, b = -0.015 }, { r = -0.009, g = 0, b = 0 } },
		levels = { 25, 25, 50, 100 },
		base_speeds = { 1.5, 2.25, 3, 3.75 },
		speed_multipliers = { 0.03, 0.03, 0.015, 0.0375 },
		base_consumption = { 125, 150, 200, 300 },
		consumption_multipliers = { 1, 2, 2, 3 },
		consumption_unit = { "kW", "kW", "kW", "kW" },
		base_pollution = { 1.8, 2.4, 3, 3.6 },
		pollution_multipliers = { 0.1, 0.1, 0.1, 0.2 },
		base_productivity = { 0, 0, 0, 0 },
		productivity_multipliers = { 0.002, 0.002, 0.002, 0.002 },
		levels_per_module_slots = { 25, 25, 25, 25 },
		base_module_slots = { 1, 2, 3, 4 },
		bonus_module_slots = { 0, 0, 1, 1 }
	}

	crystallizer_levels = {
		type = "assembling-machine",
		tiers = 2,
		base_machine_names = { "crystallizer", "crystallizer-2" },
		base_level_tints = { { r = 1, g = 1, b = 1 }, { r = 1, g = 1, b = 1 } },
		level_tint_multipliers = { { r = 0, g = 0, b = -0.018 }, { r = 0, g = 0, b = -0.009 } },
		levels = { 50, 100 },
		base_speeds = { 1.5, 2.25 },
		speed_multipliers = { 0.015, 0.0275 },
		base_consumption = { 150, 250 },
		consumption_multipliers = { 2, 3 },
		consumption_unit = { "kW", "kW", "kW" },
		base_pollution = { 1.8, 2.4 },
		pollution_multipliers = { 0.1, 0.1 },
		base_productivity = { 0, 0 },
		productivity_multipliers = { 0.002, 0.002 },
		levels_per_module_slots = { 25, 25 },
		base_module_slots = { 1, 2 },
		bonus_module_slots = { 0, 1 }
	}

	factory_levels.create_leveled_machines(ore_crusher_levels)
	factory_levels.create_leveled_machines(liquifier_levels)
	factory_levels.create_leveled_machines(crystallizer_levels)
end

